tests/lang/s02_core/s02_procedures/t03_overload_core.nim

```diff
@@ -174,12 +177,34 @@ block generic_vs_subtype:
     B = ref object of A
     C = ref object of B
 
-  proc impl[T: A](a: T): string = "matched generic"
-  proc impl(b: B): string = "matched subtype B"
-  proc impl(a: A): string = "matched subtype A"
+  block regular_procs:
+    proc impl[T: A](a: T): string = "matched generic"
+    proc impl(b: B): string = "matched subtype B"
+
+    ## subtype vs generic - wins generic
+    doAssert impl(C()) == "matched generic"
+
+    ## generic vs exact match - wins exact match
+    doAssert impl(B()) == "matched subtype B"
+
+block subtype_constaints_for_generic:
+  ## When subtype is used as a constraint for the generic parameter it
+  ## functions the same way as if it was written directly in the argument.
+
+  type
+    A  = ref object of RootObj
+    B1 = ref object of A
+    B2 = ref object of A
+
+  proc impl[T](a: T): string = "T"
+  proc impl[T: A](a: T): string = "A"
+  proc impl[T: B1](a: T): string = "B1"
+  proc impl[T: B2](a: T): string = "B2"
+  proc impl[T: B1 | B2](a: T): string = "B1|B2"
+
+  doAssert impl(B1()) == "B1"
+  doAssert impl(B2()) == "B2"
 
-  ## subtype vs generic - wins generic
-  doAssert impl(C()) == "matched generic"
+  ## Note that `B1 | B2` is not reported as an ambiguous overload since constaints
+  ## with alternatives have a lower precedence compared to the simple `SomeType`.
```

Comment:
Does this part make any sense wrt. to the implementation, or it is supposed to be an redefinition error?

Reply:
Expanded from https://github.com/nim-lang/Nim/issues/6526

Reply:
That should be a redefinition error and the more specific and concrete the type is the more it should win.

The exception being that untyped and typed happen earlier and so they're special and should happen first.